### PR TITLE
Migrate scripts and docs from tfds pipeline to grain

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

This PR migrates MaxText end-to-end integration test scripts, convergence benchmarks, DiLoCo scripts, and model documentation from the legacy `tfds` dataset pipeline to `dataset_type=grain` with `grain_file_type=tfrecord`, as a follow up of #4949 

This is a prerequisite toward removing `tensorflow` and `tensorflow_datasets` dependencies in MaxText and set the default `dataset_type=synthetic` (coming PR).

b/555209521

---

### Key Changes

1. **End-to-End Test Scripts (TPU & GPU)**
   - Updated integration test scripts across all model families (Llama 2/3/3.1/3.3, Gemma 1/2/3/4, GPT-OSS 20B/120B, Qwen 3/Next/MoE, DeepSeek V2, Mixtral) to use `dataset_type=grain grain_file_type=tfrecord dataset_path=${DATASET_PATH}`.
   - In some cases I change from synthetic data to real data, this is to ensured consistency within multi-step test scripts so all runs requiring real training data consistently use Grain TFRecord.

2. **1B Parameter Convergence Test (`test_convergence_1b_params.sh`)**
   - Migrated default training data pipeline from `tfds` to `dataset_type=grain grain_file_type=tfrecord`.
   - Not directly related to this feature, but makes testing easier: enabled overriding configuration parameters (`STEPS`, `EVAL_STEPS`, `EVAL_INTERVAL`, `PER_DEVICE_BATCH_SIZE`, `LOSS_THRESHOLD`, `DATASET_TYPE`) via environment variables / command-line arguments.
   - Removed legacy hardcoded `LIBTPU_INIT_ARGS`, which causes`FAILED_PRECONDITION` errors on some TPU generations (e.g. TPU v5p / Viperfish).

3. **Convergence Benchmarks & DiLoCo**
   - Updated `benchmarks/convergence/c4_exp.py` and `benchmarks/convergence/convergence_utils.py` to propagate `grain_file_type=tfrecord` for `c4_en_hp` and `c4_mutil_hp`.
   - Updated DiLoCo streaming and distributed training scripts (`scripts/diloco/run_diloco.sh`, `run_spmd_streaming_diloco.sh`, `run_spmd_streaming_dlco_moe_c4.sh`).

4. **Documentation & Tutorials**
   - Updated runnable command examples across model docs (`Run_Gemma3.md`, `Run_Gemma4.md`, `run_gpt_oss.md`, `Run_DeepSeek.md`, `run_qwen3_next.md`, `Run_Kimi.md`, `full_finetuning.md`) to reflect `dataset_type=grain grain_file_type=tfrecord`.

5. **Cleanup of Obsolete Tests**
   - Removed obsolete TFDS<->Grain input-pipeline checkpoint compatibility tests (`tests/end_to_end/test_checkpoint_compatibility.sh`, `tests/integration/checkpoint_compatibility_test.py`).

---

# Tests

- **XPK E2E Workload Testing (TPU v5p-8 / `cloud-tpu-multipod-dev`)**:
  - `conv1b`: Verified end-to-end execution of `test_convergence_1b_params.sh` with Grain TFRecord data loading and loss convergence verification. [cloud-log](https://cloudlogging.app.goo.gl/amZRDUJVQZCniHtG8)
  - `gemma-e2e`: Verified checkpoint conversion, fine-tuning, and pre-training runs with Grain TFRecord. [cloud-log](https://cloudlogging.app.goo.gl/inYLFtNKwehdXT7i7). The script failed at later forward_logit_checker step for loading tokenizer, which is a pre-existing bug unrelated to this PR

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: tests/end_to_end/tpu/gemma/2b/test_gemma.sh [hunk 1]
```diff
@@ -46,7 +46,7 @@ python3 -m maxtext.inference.decode "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:
...
 export FINETUNE_RUN_NAME=runner_finetune_${idx}
-python3 -m maxtext.trainers.pre_train.train "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/configs}"/base.yml base_output_directory=${BASE_OUTPUT_DIRECTORY} dataset_path=${DATASET_PATH} tokenizer_path="${MAXTEXT_ASSETS_ROOT:-${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/assets/tokenizers}}"/tokenizer.gemma load_parameters_path=${CONVERTED_CHECKPOINT} per_device_batch_size=1 run_name=${FINETUNE_RUN_NAME} max_target_length=8192 steps=10 async_checkpointing=false model_name=gemma-2b checkpoint_period=5
+python3 -m maxtext.trainers.pre_train.train "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/configs}"/base.yml base_output_directory=${BASE_OUTPUT_DIRECTORY} dataset_type=grain grain_file_type=tfrecord dataset_path=${DATASET_PATH} tokenizer_path="${MAXTEXT_ASSETS_ROOT:-${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/assets/tokenizers}}"/tokenizer.gemma load_parameters_path=${CONVERTED_CHECKPOINT} per_device_batch_size=1 run_name=${FINETUNE_RUN_NAME} max_target_length=8192 steps=10 async_checkpointing=false model_name=gemma-2b checkpoint_period=5
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tests/end_to_end/tpu/gemma/2b/test_gemma.sh:49]: done

File: tests/end_to_end/tpu/test_convergence_1b_params.sh [hunk 2]
```diff
@@ -13,14 +13,14 @@ echo "Running test_convergence_1b_params.sh"
...
-export EVAL_INTERVAL=100
-export DATASET_TYPE=grain
-export MTP_NUM_LAYERS=0 # Disable MTP by default
-export PER_DEVICE_BATCH_SIZE=8.0 # With the default learning rate (3e-4) this should have global batch of 512, with 2k sequence length (1M global batch in tokens)
+# default values, can be overridden from environment variables or command line
+export LOSS_THRESHOLD="${LOSS_THRESHOLD:-100.0}" # Set to large value so test is guaranteed to pass.
+export STEPS="${STEPS:-20400}" # Run for 20B tokens for a 1B sized mode for "chinchilla" scaling https://arxiv.org/abs/2203.15556
+export EVAL_STEPS="${EVAL_STEPS:-160}"
+export EVAL_INTERVAL="${EVAL_INTERVAL:-100}"
+export DATASET_TYPE="${DATASET_TYPE:-""}" # Default: grain tfrecord
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tests/end_to_end/tpu/test_convergence_1b_params.sh]: done

File: tests/end_to_end/tpu/test_convergence_1b_params.sh [hunk 3]
```diff
@@ -33,24 +33,26 @@ then
     export M_RUN_NAME=$RUN_NAME
 fi
 
-if [ "$DATASET_TYPE" == "grain" ]
+if [ -z "$DATASET_TYPE" ] || [ "$DATASET_TYPE" == "grain_tfrecord" ]
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tests/end_to_end/tpu/test_convergence_1b_params.sh]: done

File: tests/end_to_end/tpu/llama2/7b/test_llama2_7b.sh [hunk 4]
```diff
@@ -55,7 +55,7 @@ python3 -m maxtext.inference.decode "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:
...
 # We run this with context parallelism and the `ici_context_parallelism` flag as an integration test for context parallelism
-python3 -m maxtext.trainers.pre_train.train "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/configs}"/base.yml load_parameters_path=${CONVERTED_CHECKPOINT} run_name=runner_finetuning_${idx} base_output_directory=${BASE_OUTPUT_DIRECTORY} dataset_path=${DATASET_PATH} async_checkpointing=${ASYNC_CHECKPOINTING} per_device_batch_size=1 model_name='llama2-7b' ici_context_parallelism=4 steps=10 per_device_batch_size=1 checkpoint_period=5 packing=false
+python3 -m maxtext.trainers.pre_train.train "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/configs}"/base.yml load_parameters_path=${CONVERTED_CHECKPOINT} run_name=runner_finetuning_${idx} base_output_directory=${BASE_OUTPUT_DIRECTORY} dataset_type=grain grain_file_type=tfrecord dataset_path=${DATASET_PATH} async_checkpointing=${ASYNC_CHECKPOINTING} per_device_batch_size=1 model_name='llama2-7b' ici_context_parallelism=4 steps=10 per_device_batch_size=1 checkpoint_period=5 packing=false
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tests/end_to_end/tpu/llama2/7b/test_llama2_7b.sh]: done

File: tests/end_to_end/tpu/llama_finetuning_test.sh [hunk 5]
```diff
@@ -13,7 +13,7 @@ DATASET_PATH=gs://maxtext-dataset
 
 export LOSS_THRESHOLD=2.5
 
-python3 -m maxtext.trainers.pre_train.train "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/configs}"/base.yml run_name=runner_direct_${idx} base_output_directory=${BASE_OUTPUT_DIRECTORY} load_parameters_path=${base_ckpt_path} model_name='llama2-7b' dataset_path=${DATASET_PATH} async_checkpointing=false  model_name='llama2-7b' ici_tensor_parallelism=4 steps=10 per_device_batch_size=.25 metrics_file='metrics.txt'
+python3 -m maxtext.trainers.pre_train.train "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/configs}"/base.yml run_name=runner_direct_${idx} base_output_directory=${BASE_OUTPUT_DIRECTORY} load_parameters_path=${base_ckpt_path} model_name='llama2-7b' dataset_type=grain grain_file_type=tfrecord dataset_path=${DATASET_PATH} async_checkpointing=false  model_name='llama2-7b' ici_tensor_parallelism=4 steps=10 per_device_batch_size=.25 metrics_file='metrics.txt'
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tests/end_to_end/tpu/llama_finetuning_test.sh]: done

File: tests/end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh [hunk 6]
```diff
@@ -65,10 +65,10 @@ fi
...
 
 # Run pre-training - tokamax_gmm implementation
-python3 -m maxtext.trainers.pre_train.train "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/configs}"//base.yml base_output_directory=${BASE_OUTPUT_PATH} run_name=tokamax_gmm_pre_training model_name=${MODEL_NAME} tokenizer_type=huggingface tokenizer_path=${TOKENIZER_PATH} dataset_type=synthetic enable_checkpointing=false attention=flash sparse_matmul=True use_tokamax_gmm=True dtype=bfloat16 weight_dtype=bfloat16 per_device_batch_size=4 steps=5 max_target_length=1024 ici_fsdp_parallelism=4
+python3 -m maxtext.trainers.pre_train.train "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/configs}"//base.yml base_output_directory=${BASE_OUTPUT_PATH} run_name=tokamax_gmm_pre_training model_name=${MODEL_NAME} tokenizer_type=huggingface tokenizer_path=${TOKENIZER_PATH} dataset_type=grain grain_file_type=tfrecord dataset_path=${DATASET_PATH} enable_checkpointing=false attention=flash sparse_matmul=True use_tokamax_gmm=True dtype=bfloat16 weight_dtype=bfloat16 per_device_batch_size=4 steps=5 max_target_length=1024 ici_fsdp_parallelism=4
```
Comment (aireenmei) [inline tests/end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh:68]: I update this from synthetic to real data to be consistent with the later "fine-tuning" section, which requires real data for testing decode later

**Review by SurbhiJainUSC — APPROVED** [review]

**Review by bvandermoon — APPROVED** [review]
